### PR TITLE
[sca] Add initial implementation of SHA3 SCA program

### PR DESCRIPTION
**Pull request by alphan**
This PR adds a program for SHA3/KMAC SCA. I also improved `meson.build` a bit and refactored the code for trigger selection and disabling peripherals, which will hopefully make it easier to add new programs in the future. I'll create a separate PR in lowRISC/ot-sca for the capture scripts. I also had to write some static functions for controlling KMAC to unblock this PR but these can be moved to the KMAC DIF library later.

I'm not sure what the attack(s) will look like but I guess we could approach this in two ways:
1. Performing a `SHA3(key | message)` operation (@moidx's suggestion), or
2. Loading the key to KMAC and performing a `KMAC` operation.

I went with the latter to exclude key loads from traces. That being said, (1) can also be easily implemented as a separate command handler if needed.

This change does not include a batch capture command and the current capture speed is ~100 traces/s. This is how traces look like:
Screen Shot 2021-08-14 at 00 09 51

The trigger is high for \~48 us but the traces are a bit longer to include the idle state as well. @eunchan told me that he would expect this operation (absorb with a 16 byte key and message) to take 3\~4 us at 100 MHz. Since CW310 runs at 10 MHz and we also have some overhead due to DIFs, the length of the trigger seems to match our expectations.

Another thing to note is that since the KMAC operations require the CPU for FIFO writes, we cannot use `sca_call_and_sleep` as we did for AES and traces will have some additional noise as a result. I'm not sure if we can do something in HW to improve this for SCA.

Also, we have been calling this `sha3_serial` so far but maybe we should call it `kmac_serial` since it uses the KMAC peripheral. Please let me know what you think.

### CODE REVIEW

File: sw/device/sca/sha3_serial.c
Comment (mundaym) [inline sw/device/sca/sha3_serial.c:144]: This KMAC operation won't return the right result because it doesn't append the output length (d) to the message (this is done as part of `dif_kmac_squeeze`). Maybe worth a comment in case someone uses this code to try and calculate the actual KMAC value.

File: sw/device/sca/sha3_serial.c
Comment (mundaym) [inline sw/device/sca/sha3_serial.c:145]: Do you agree we could drop this call if `dif_kmac_squeeze(&kmac, NULL, 0, NULL)` blocked until the squeeze state was entered? (Just thinking how we could make the DIF support this use case.)

File: sw/device/sca/sha3_serial.c
Comment (mundaym) [inline sw/device/sca/sha3_serial.c:105]: I don't think you need this. `dif_kmac_mode_kmac_start` will block until KMAC is idle (which it should be since `dif_kmac_configure` also blocks until in this state).

**Review by vogelpi — APPROVED** [review]
Thanks @alphan for working on this! Especially the changes to the shared SCA code look good to me. About the kmac/sha3-specific code: I am not an expert here as I don't know the design very well. But having a look at the DIF and hardware documentation, the following came to my mind:
- The hardware IP should trigger a "done" interrupt after the absorb step. So it may be possible to enter sleep after providing the message with `dif_kmac_absorb` and prevent the polling.
- The message FIFO in hardware is 64 entries deep and thus sufficiently big to hold the entire message for this use case. But I don't know if the absorb already starts when writing the first message byte or only once writing the process command (currently inside `kmac_absorb_end()`). Depending on that, it may make sense to add a delay in hardware for SCA testing only as we have it for AES, too.

It would be important that also @vrozic reviews and comments on this PR as he has been thinking on attack/analysis scripts for SHA3/KMAC previously. He can probably answer your question about the strategy.

File: sw/device/sca/sha3_serial.c
Comment (alphan) [reply sw/device/sca/sha3_serial.c:144]: Yes, we are not interested in the squeeze step yet but I added note. Thanks for pointing this out.

File: sw/device/sca/sha3_serial.c
Comment (alphan) [reply sw/device/sca/sha3_serial.c:145]: I think so. But I'm not sure if the part that appends the length would overlap with another HW operation and add to the noise. Also, if the absorb can be delayed until the "process" command (as @vogelpi [commented](https://github.com/lowRISC/opentitan/pull/7757#pullrequestreview-730473857)) then we would prefer putting Ibex to sleep instead of blocking until the squeeze step.

File: sw/device/sca/sha3_serial.c
Comment (alphan) [reply sw/device/sca/sha3_serial.c:105]: Hmm, both functions seem to return errors instead of blocking if the hardware is not idle.

**Comment by alphan** [comment]
> * The hardware IP should trigger a "done" interrupt after the absorb step. So it may be possible to enter sleep after providing the message with `dif_kmac_absorb` and prevent the polling.
> * The message FIFO in hardware is 64 entries deep and thus sufficiently big to hold the entire message for this use case. But I don't know if the absorb already starts when writing the first message byte or only once writing the process command (currently inside `kmac_absorb_end()`). Depending on that, it may make sense to add a delay in hardware for SCA testing only as we have it for AES, too.

Thanks for the review @vogelpi, added to meeting notes. We should definitely discuss these with @eunchan when we start working on an attack.

File: sw/device/sca/sha3_serial.c
Comment (mundaym) [reply sw/device/sca/sha3_serial.c:105]: Sorry for the noise, you're right. Not sure what I was thinking there.

We could add a 'not idle' error to them which might be useful.

File: sw/device/sca/lib/sca.c
Comment (moidx) [inline sw/device/sca/lib/sca.c:160]: I would implement these functions directly in the difs as is. Whenever a peripheral needs to be updated, it takes time to find and update all the places that don't consume DIFs.

File: sw/device/sca/sha3_serial.c
Comment (moidx) [inline sw/device/sca/sha3_serial.c:78]: We should discuss with @eunchan to add a parameter to delay the start of the operation for testing purposes,  and a TODO reference here. Maybe this is sufficient to switch to the call and sleep approach.

**Review by moidx — APPROVED** [review]

**Comment by tjaychen** [comment]
> This PR adds a program for SHA3/KMAC SCA. I also improved `meson.build` a bit and refactored the code for trigger selection and disabling peripherals, which will hopefully make it easier to add new programs in the future. I'll create a separate PR in lowRISC/ot-sca for the capture scripts. I also had to write some static functions for controlling KMAC to unblock this PR but these can be moved to the KMAC DIF library later.
> 
> I'm not sure what the attack(s) will look like but I guess we could approach this in two ways:
> 
> 1. Performing a `SHA3(key | message)` operation (@moidx's suggestion), or
> 2. Loading the key to KMAC and performing a `KMAC` operation.
> 
> I went with the latter to exclude key loads from traces. That being said, (1) can also be easily implemented as a separate command handler if needed.
> 
> This change does not include a batch capture command and the current capture speed is ~100 traces/s. This is how traces look like:
> Screen Shot 2021-08-14 at 00 09 51
> 
> The trigger is high for ~48 us but the traces are a bit longer to include the idle state as well. @eunchan told me that he would expect this operation (absorb with a 16 byte key and message) to take 3~4 us at 100 MHz. Since CW310 runs at 10 MHz and we also have some overhead due to DIFs, the length of the trigger seems to match our expectations.
> 
> Another thing to note is that since the KMAC operations require the CPU for FIFO writes, we cannot use `sca_call_and_sleep` as we did for AES and traces will have some additional noise as a result. I'm not sure if we can do something in HW to improve this for SCA.

can you clarify this a bit? We don't have a DMA in our system, so wouldn't we need to CPU to do data moves even for AES? I didn't completely follow. 


> 
> Also, we have been calling this `sha3_serial` so far but maybe we should call it `kmac_serial` since it uses the KMAC peripheral. Please let me know what you think.

**Comment by alphan** [comment]
> can you clarify this a bit? We don't have a DMA in our system, so wouldn't we need to CPU to do data moves even for AES? I didn't completely follow.

@tjaychen: Sure, for AES, we first load the data to `AES_DATA_IN_*` registers, then start the encryption manually by writing to the `AES_TRIGGER` register. AES is configured to start 40 cycles after this which gives us enough time to put Ibex to sleep.

File: sw/device/sca/lib/sca.c
Comment (alphan) [reply sw/device/sca/lib/sca.c:160]: Added a note about `dif_end_stop()` in #5465 and created #7837 for `dif_csrng_stop()`.

File: sw/device/sca/sha3_serial.c
Comment (alphan) [reply sw/device/sca/sha3_serial.c:78]: Captured and added links to your and @vogelpi's comments in #7841 and added a TODO here.

File: sw/device/sca/sha3_serial.c
Comment (alphan) [reply sw/device/sca/sha3_serial.c:108]: Created #7842.

**Comment by alphan** [comment]
I will merge this PR after CI runs since we will probably do a few iterations while working on an attack.

**Pull request merged**
